node/txsimulator/txSimulator.go

```diff
@@ -0,0 +1,151 @@
+package txsimulator
+
+import (
+	"encoding/hex"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/data/block"
+	"github.com/ElrondNetwork/elrond-go/data/receipt"
+	"github.com/ElrondNetwork/elrond-go/data/smartContractResult"
+	"github.com/ElrondNetwork/elrond-go/data/transaction"
+	"github.com/ElrondNetwork/elrond-go/node"
+	"github.com/ElrondNetwork/elrond-go/process"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+// Args holds the arguments required for creating a new transaction simulator
+type Args struct {
+	TransactionProcessor       TransactionProcessor
+	IntermmediateProcContainer process.IntermediateProcessorContainer
+	AddressPubKeyConverter     core.PubkeyConverter
+}
+
+type transactionSimulator struct {
+	txProcessor            TransactionProcessor
+	intermProcContainer    process.IntermediateProcessorContainer
+	addressPubKeyConverter core.PubkeyConverter
+}
+
+// New returns a new instance of a transactionSimulator
+func New(args Args) (*transactionSimulator, error) {
+	if check.IfNil(args.TransactionProcessor) {
+		return nil, node.ErrNilTxSimulatorProcessor
+	}
+	if check.IfNil(args.IntermmediateProcContainer) {
+		return nil, node.ErrNilIntermediateProcessorContainer
+	}
+	if check.IfNil(args.AddressPubKeyConverter) {
+		return nil, node.ErrNilPubkeyConverter
+	}
+
+	return &transactionSimulator{
+		txProcessor:            args.TransactionProcessor,
+		intermProcContainer:    args.IntermmediateProcContainer,
+		addressPubKeyConverter: args.AddressPubKeyConverter,
+	}, nil
+}
+
+// ProcessTx will process the transaction in a special environment, where state-writing is not allowed
+func (ts *transactionSimulator) ProcessTx(tx *transaction.Transaction) (*transaction.SimulationResults, error) {
+	txStatus := core.TxStatusReceived
+	failReason := ""
+	retCode, err := ts.txProcessor.ProcessTransaction(tx)
+	if err != nil {
+		failReason = err.Error()
+		txStatus = core.TxStatusNotExecuted
+	}
+
+	if retCode == vmcommon.Ok {
+		txStatus = core.TxStatusExecuted
+	}
+
+	results := &transaction.SimulationResults{
+		Status:     txStatus,
+		FailReason: failReason,
+	}
+
+	err = ts.addIntermediateTxsToResult(results)
+	if err != nil {
+		return nil, err
+	}
+
+	return results, nil
+}
+
+func (ts *transactionSimulator) addIntermediateTxsToResult(result *transaction.SimulationResults) error {
+	scrForwarder, err := ts.intermProcContainer.Get(block.SmartContractResultBlock)
+	if err != nil {
+		return err
+	}
+	receiptsForwarder, err := ts.intermProcContainer.Get(block.ReceiptBlock)
```

Comment:
there is no receipt block in case of metachain - thus this will fail.
What you should do it iterate all processors and call createblockstarted. Receipts are created only in intra shard move balance transfers with too much gas.

Reply:
done. in case of meta, apply only the smart contract results and in case of shard, also the receipts. and on defer now it closes all the processors